controlplane/kubeadm/controllers/helpers.go

```diff
@@ -246,3 +247,24 @@ func (r *KubeadmControlPlaneReconciler) generateMachine(ctx context.Context, kcp
 	}
 	return nil
 }
+
+// MachinesNeedingRollout return a list of machines that need to be upgraded.
+func (r *KubeadmControlPlaneReconciler) MachinesNeedingRollout(ctx context.Context, c *internal.ControlPlane) internal.FilterableMachineCollection {
+	now := metav1.Now()
+
+	// Ignore machines to be deleted.
+	machinesToUpgrade := c.Machines.Filter(machinefilters.Not(machinefilters.HasDeletionTimestamp))
```

Comment:
I'm wondering if we should short-circuit logic in `reconcile()` to return early if there are any owned machines that are in the middle of deletion, I can't think of any actions outside of status update that we would want to take while there is a Machine that is in the process of deletion. I would say that we wouldn't even want to update observedGeneration in this case, since we are not doing anything to reconcile towards the desired spec.

Reply:
I think we can do it in a separate PR after MHC remediation is in, I remember seeing those checks in that PR.

Reply:
Probably fine to leave this for now, the remediation PR should take care of this.